caffe2/sgd/adadelta_op.h

```diff
@@ -0,0 +1,179 @@
+    adadelta_update<Context>(
+        Input(GRAD).size(),
+        Input(PARAM).template data<T>(),
+        Input(GRAD).template data<T>(),
+        Input(MOMENT_GRAD).template data<T>(),
+        Input(MOMENT_DELTA).template data<T>(),
+        Output(OUTPUT_PARAM)->template mutable_data<T>(),
+        Output(OUTPUT_MOMENT_GRAD)->template mutable_data<T>(),
+        Output(OUTPUT_MOMENT_DELTA)->template mutable_data<T>(),
+        epsilon_,
+        decay_,
+        Input(LR).template data<T>(),
+        &context_);
+    return true;
+  }
+
+ protected:
+  T epsilon_;
+  T decay_;
+  INPUT_TAGS(PARAM, MOMENT_GRAD, MOMENT_DELTA, GRAD, LR);
+  OUTPUT_TAGS(OUTPUT_PARAM, OUTPUT_MOMENT_GRAD, OUTPUT_MOMENT_DELTA);
+};
+
+template <typename T, class Context>
+class SparseAdadeltaOp final : public Operator<Context> {
+ public:
+  USE_OPERATOR_CONTEXT_FUNCTIONS;
+  SparseAdadeltaOp(const OperatorDef& operator_def, Workspace* ws)
+      : Operator<Context>(operator_def, ws),
+        epsilon_(OperatorBase::GetSingleArgument<T>("epsilon", 1e-5f)),
+        decay_(OperatorBase::GetSingleArgument<T>("decay", 0.95f)) {}
+
+  bool RunOnDevice() override {
+    // Enforce shapes
+    CAFFE_ENFORCE_EQ(Input(PARAM).size(), Input(MOMENT_GRAD).size());
+    CAFFE_ENFORCE_EQ(Input(PARAM).size(), Input(MOMENT_DELTA).size());
+    CAFFE_ENFORCE_EQ(Input(LR).size(), 1);
+    CAFFE_ENFORCE_EQ(
+        Input(PARAM).size_from_dim(1),
+        Input(GRAD).size_from_dim(Input(INDICES).ndim()));
+    // Enforce domain constraints on attributes
+    CAFFE_ENFORCE_GE(static_cast<T>(epsilon_), 0.0f);
+    CAFFE_ENFORCE_GT(static_cast<T>(decay_), 0.0f);
+    CAFFE_ENFORCE_LT(static_cast<T>(decay_), 1.0f);
+
+    return DispatchHelper<TensorTypes<int32_t, int64_t>>::call(
+        this, Input(INDICES));
+  }
+
+  template <typename SIndex>
+  bool DoRunWithType() {
+    const auto* lr = Input(LR).template data<T>();
+    const auto* indices = Input(INDICES).template data<SIndex>();
+    const auto* gradIn = Input(GRAD).template data<T>();
+    const auto* paramIn = Input(PARAM).template data<T>();
+    const auto* momentIn = Input(MOMENT_GRAD).template data<T>();
+    const auto* momentDeltaIn = Input(MOMENT_DELTA).template data<T>();
+    auto* paramOut = Output(OUTPUT_PARAM)->template mutable_data<T>();
+    auto* momentOut = Output(OUTPUT_MOMENT_GRAD)->template mutable_data<T>();
+    auto* momentDeltaOut =
+        Output(OUTPUT_MOMENT_DELTA)->template mutable_data<T>();
+
+    auto n = Input(INDICES).size();
+    if (n == 0) {
+      return true;
+    }
+
+    auto block_size = Input(GRAD).size() / n;
+    for (auto i = 0; i < n; ++i) {
+      auto idx = indices[i];
+      if (block_size == 1) {
+        float gi = gradIn[i];
+        float di = momentDeltaIn[idx];
+        float hi = momentOut[idx] =
+            decay_ * momentIn[idx] + (static_cast<T>(1) - decay_) * gi * gi;
+        float ng = (std::sqrt(di + epsilon_) / std::sqrt(hi + epsilon_)) * gi;
```

Comment:
use std::rsqrt here . std::sqrt is slower than it and actually there is no need to use such divide here.